### ISSUE TITLE
Feature request: replicate the native MacOS Aerial screensaver->wallpaper transition

### ISSUE THREAD

**Issue by rmyadsk** [JohnCoates/Aerial#1376]
- [ ] Mac model: MacBook Air M2 (2022)
- [ ] macOS version: Sequoia 15.1 Beta 3
- [ ] Monitor setup: Apple Studio Display with laptop clamshell closed while in home office / built-in Air display while mobile

Feature request here. recent MacOS distributions have an option for the screensaver to essentially become the wallpaper. In the native MacOS mode, as the source files appear to be high-res .mov files playing on loop, the screensaver plays at what feels like rabbit speed playback on this project. Then, when a user interrupts the screensaver, the scene gradually+incrementally+gracefully slows and ultimately freezes in the space of a couple seconds mid-scene.

**Comment by glouel** [comment]
Yep it's something I'm working on. There are a few things I need to solve though for multiple space/monitor support but it's planned !

**Comment by glouel** [comment]
Took a while but it's here, we have continuity, speed transition (speeds up to 1x on screensaver, slows down to your preset desktop speed if it's running slower) AND auto pause based on window coverage :)  

https://github.com/AerialScreensaver/Aerial/releases/tag/v3.9.9a27

**Comment by rmyadsk** [comment]
@glouel Thanks for the release! Definitely worth upgrading. A few notes on v3.9.9a27:

1. There's an interesting artifact when in live wallpaper mode (possibly in screensaver too?) - when coming to the end of a video, the screen will fade to black, then unfade and show about a half-second or so of video, then fade to black again before switching to the next video.
2. The lock screen functionality works when locking from the Aerial control applet, but not from the OS-wide Apple icon -> Lock Screen. When locking via the latter method I get the same behavior as if Aerial was exited and then I lock the screen - my stock wallpaper and a clock.
3. It might be convenient for the user to customize what percentage of the screen needs to be covered by windows in order to pause the wallpaper movement. 
4. Tooltip texts for the buttons/icons/controls in the applet would be useful.

Misc:
1. I have this loaded on a M2 MacBook and an M4 Mac Mini. Functions as expected on both.
2. The "Display" detection works perfectly.
3. The one-time migration process, despite having a large cache, worked almost instantly for me.
4. Having the Aerial Companion functionality built-in to the release via the applet is great. As far as a I can tell Aerial Companion can be safely uninstalled after this release is loaded.

Great job! I can't wait to watch this progress through dev and test into production.

**Comment by glouel** [comment]
> 1. There's an interesting artifact when in live wallpaper mode (possibly in screensaver too?) - when coming to the end of a video, the screen will fade to black, then unfade and show about a half-second or so of video, then fade to black again before switching to the next video.

Hmm I was pretty sure I had fixed that one, I'll double check thanks ! If you go into the app settings (2nd icon bottom left in the menubar popup) and go to advanced, what do you have set here : 

<img width="2124" height="1728" alt="Image" src="https://github.com/user-attachments/assets/66b8009e-6baa-4a22-8219-5ae7f879267b" />

Also which format do you have on top ? It may not match the one you had previously, I think I have a bug around that during migration, would be good to know while you're there. If it doesn't match, don't change it yet.

Historically I had the fades as a way to hide the video loading, but it's a non issue nowadays. I'm not sure I like the fades on desktop mode at this point, they are fine on screensaver but I'm a bit bothered by them on desktop. I may need to split that setting.

> 2. The lock screen functionality works when locking from the Aerial control applet, but not from the OS-wide Apple icon -> Lock Screen. When locking via the latter method I get the same behavior as if Aerial was exited and then I lock the screen - my stock wallpaper and a clock.

So maybe to explain exactly how it works, there are two parts to this : 
- Aerial.app : that's what you call the applet, it's the app in your menu bar that does the video desktop feature. MacOS doesn't natively support video playback on desktop, so that's what the app does (basically, it draws a video on top of your fixed wallpaper, and you can see where this is going).  
- AerialScreenSaverExtension : that's the screensaver part (using an app extension)   

When you press lock screen in the app, I start the screensaver. While it's running, if you move your mouse/keyboard, macOS will put the login UI on top (I hide my overlays then) and let you log in. 

When you press lock screen in macOS, macOS just locks your screen with your desktop background, which is not my video, and will go to sleep after a bit. You'll have the same scenario if you put your mac to sleep and wake it up, macos will always show your desktop background. 

The only way for me to give a better sense of continuity would be to generate on the fly a screenshot of the video and put it as your desktop background. It's on my list of things to try  but there may be timing issues. I'm not sure I can generate it and replace it fast enough 🤷

> 4. It might be convenient for the user to customize what percentage of the screen needs to be covered by windows in order to pause the wallpaper movement.

It may be a bit hidden in settings but it's there, you can test it live : 

<img width="2124" height="1728" alt="Image" src="https://github.com/user-attachments/assets/1fd3858a-b4a7-4adf-a4bb-9b2c147d5ad0" />

> 6. Tooltip texts for the buttons/icons/controls in the applet would be useful.

Yep that's on my to do list, there are a few but a bunch are missing. 

> 1. I have this loaded on a M2 MacBook and an M4 Mac Mini. Functions as expected on both.

Great

> 2. The "Display" detection works perfectly.

Out of curiosity do you have multiple monitors with the same resolution ? 

If you do, you'll want Advanced screen detection on, but it adds a bit of a slow down when screensaver starts (on latest versions I have a spinner for this)

If you don't, uncheck advanced screen detection for an instant start of the screensaver. It's default off now in latest builds.

> 4. The one-time migration process, despite having a large cache, worked almost instantly for me.

💪

> 5. Having the Aerial Companion functionality built-in to the release via the applet is great. As far as a I can tell Aerial Companion can be safely uninstalled after this release is loaded.

Yes it's basically the replacement for Companion, I renamed it Aerial.app since it has both the app and screensaver inside. A lot of features depend on the app running, as that's the only way I can go around a bunch of restrictions on what the screensaver is allowed to do by macOS.  

> Great job! I can't wait to watch this progress through dev and test into production.

Thanks for your feedback, there still needs a bit of polish but it's getting close. Let me know if you see more issues, also make sure you check the auto updates, there's been a couple of builds since, but none I think that would affect what you reported.

**Comment by glouel** [comment]
<img width="1149" height="1199" alt="Image" src="https://github.com/user-attachments/assets/afe37ef1-3ba5-407f-9687-50c2d0d5368f" />

So V4 is out and I've added a wallpaper continuity feature where I manipulate the system wallpaper, this allow for a better transition on lock screen, you get the paused video on lock screen now instead of your wallpaper. V4 is available here : https://aerialscreensaver.github.io
